### PR TITLE
ADR11: How to avoid backward incompatible migrations

### PR DESCRIPTION
**Pull request by pahaz**

### CODE REVIEW

File: apps/condo/bin/fix-database-migration-state.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+const { prepareKeystoneExpressApp } = require('@core/keystone/test.utils')
+
+const index = require('@app/condo')
+
+function exit (code, error) {
+    console.error(error)
+    process.exit(code)
+}
+
+function result (obj) {
```
Comment (akarjakin) [inline apps/condo/bin/fix-database-migration-state.js:19]: logResult

**Review by akarjakin — APPROVED** [review]

**Comment by AntonAL** [comment]
It's unclear from file name `fix-database-migration-state.js` what it does.
Maybe to rename it to `drop-orphaned-database-objects-after-dangerous-migrations.js`

File: apps/condo/bin/fix-database-migration-state.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
+/**
+ * We don't remove any tables/columns from the database for backward compatibility!
+ * But we same such tables/columns here to apply it after some day!
+ * This script just remove legacy backward compatibility tables and columns!
```
Comment (toplenboren) [inline apps/condo/bin/fix-database-migration-state.js:4]:
```suggestion
 * This script simply removes legacy backward compatibility tables and columns!
```

File: apps/condo/bin/fix-database-migration-state.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
+/**
+ * We don't remove any tables/columns from the database for backward compatibility!
+ * But we same such tables/columns here to apply it after some day!
```
Comment (toplenboren) [inline apps/condo/bin/fix-database-migration-state.js:3]: `we same such tables` doesnt look right

File: apps/condo/bin/fix-database-migration-state.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    console.info('[INFO] Init keystone ...')
+
+    console.time('init')
+    const { knex } = await init()
+    console.timeEnd('init')
+
+    return result(await knex.raw(`
+BEGIN;
+--
+-- We don't remove any tables/columns from the database for backward compatibility! But we same such tables/columns here to apply it after some day!
```
Comment (toplenboren) [inline apps/condo/bin/fix-database-migration-state.js:51]: so we store changes here?

**Review by toplenboren — APPROVED** [review]

File: apps/condo/bin/fix-database-migration-state.js (hunk 3 above)
Comment (pahaz) [inline apps/condo/bin/fix-database-migration-state.js:3]:
```suggestion
 * But we need to save backward-incompatible changes to apply it someday! 
 * Backward compatibility is required but we also need a way to remove legacy compatibility tables and columns.
 * This script solves this problem!
```

File: apps/condo/bin/fix-database-migration-state.js (hunk 1 above)
Comment (pahaz) [inline apps/condo/bin/fix-database-migration-state.js:19]:
```suggestion
function printResult (obj) {
```

File: apps/condo/bin/fix-database-migration-state.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+async function main () {
+    console.info('[INFO] Init keystone ...')
+
+    console.time('init')
+    const { knex } = await init()
+    console.timeEnd('init')
+
+    return result(await knex.raw(`
```
Comment (pahaz) [inline apps/condo/bin/fix-database-migration-state.js:48]:
```suggestion
    return printResult(await knex.raw(`
```

**Pull request merged**